File: chrome/browser/ui/android/autofill/internal/java/src/org/chromium/chrome/browser/ui/autofill/ephemeraltab/PaymentsWindowBridge.java
// Copyright 2025 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.ui.autofill.ephemeraltab;

import org.jni_zero.CalledByNative;
import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.build.annotations.NullMarked;
import org.chromium.content_public.browser.WebContents;
import org.chromium.url.GURL;

/** Bridge for the Emphemeral Tab bottom sheet. */
@JNINamespace("autofill::payments")
@NullMarked
class PaymentsWindowBridge {
    private long mNativePaymentsWindowBridge;
    private PaymentsWindowCoordinator mPaymentsWindowCoordinator;

    @CalledByNative
    PaymentsWindowBridge(long nativePaymentsWindowBridge) {
        mNativePaymentsWindowBridge = nativePaymentsWindowBridge;
        mPaymentsWindowCoordinator = new PaymentsWindowCoordinator(this);
    }

    @CalledByNative
    public void openEphemeralTab(GURL url, String title, WebContents merchantWebContents) {
        mPaymentsWindowCoordinator.openEphemeralTab(url, title, merchantWebContents);
    }

    @CalledByNative
    public void closeEphemeralTab() {
        mPaymentsWindowCoordinator.closeEphemeralTab();
    }

    @CalledByNative
    public void onNativeDestroyed() {
        mNativePaymentsWindowBridge = 0;
    }

    PaymentsWindowCoordinator getPaymentsWindowCoordinatorForTesting() {
        return mPaymentsWindowCoordinator;
    }

    void setPaymentsWindowCoordinatorForTesting(
            PaymentsWindowCoordinator paymentsWindowCoordinator) {
        mPaymentsWindowCoordinator = paymentsWindowCoordinator;
    }

    /**
     * Called when the current navigation to a page has finished.
     *
     * @param clickedUrl The URL that the user initiated the navigation to.
     */
    void onNavigationFinished(GURL clickedUrl) {
        if (mNativePaymentsWindowBridge == 0) return;
        PaymentsWindowBridgeJni.get().onNavigationFinished(mNativePaymentsWindowBridge, clickedUrl);
    }

    /**
     * Called when the observation of the WebContents has started.
     *
     * @param webContents The WebContents that is being observed.
     */
    public void onWebContentsObservationStarted(WebContents webContents) {
        if (mNativePaymentsWindowBridge == 0) return;
        PaymentsWindowBridgeJni.get()
                .onWebContentsObservationStarted(mNativePaymentsWindowBridge, webContents);
    }

    /**
     * Called when the observed {@code WebContents} is being destroyed.
     *
     * <p>After this call, clients should assume that the observed {@code WebContents} will be
     * imminently destroyed.
     */
    void onWebContentsDestroyed() {
        if (mNativePaymentsWindowBridge == 0) return;
        PaymentsWindowBridgeJni.get().onWebContentsDestroyed(mNativePaymentsWindowBridge);
    }

    void onUserDeniedTabOpening() {
        if (mNativePaymentsWindowBridge == 0) return;
        PaymentsWindowBridgeJni.get().onUserDeniedTabOpening(mNativePaymentsWindowBridge);
    }

    @NativeMethods
    interface Natives {
        void onNavigationFinished(long nativePaymentsWindowBridge, GURL clickedUrl);

        void onWebContentsObservationStarted(
                long nativePaymentsWindowBridge, WebContents webContents);

        void onWebContentsDestroyed(long nativePaymentsWindowBridge);

        void onUserDeniedTabOpening(long nativePaymentsWindowBridge);
    }
}

File: chrome/browser/ui/android/autofill/payments/payments_window_bridge.cc
// Copyright 2025 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/ui/android/autofill/payments/payments_window_bridge.h"

#include <string>

#include "base/android/jni_android.h"
#include "base/android/jni_string.h"
#include "base/check_deref.h"
#include "chrome/browser/ui/android/autofill/payments/payments_window_delegate.h"
#include "content/public/browser/web_contents.h"
#include "url/android/gurl_android.h"
#include "url/gurl.h"

// Must come after headers that provide symbols used by @JniType.
#include "chrome/browser/ui/android/autofill/internal/jni_headers/PaymentsWindowBridge_jni.h"

namespace autofill::payments {

PaymentsWindowBridge::PaymentsWindowBridge(
    PaymentsWindowDelegate* payments_window_delegate)
    : payments_window_delegate_(CHECK_DEREF(payments_window_delegate)) {
  java_payments_window_bridge_ = Java_PaymentsWindowBridge_Constructor(
      base::android::AttachCurrentThread(), reinterpret_cast<int64_t>(this));
}

PaymentsWindowBridge::~PaymentsWindowBridge() {
  if (java_payments_window_bridge_) {
    Java_PaymentsWindowBridge_onNativeDestroyed(
        base::android::AttachCurrentThread(), java_payments_window_bridge_);
  }
}

void PaymentsWindowBridge::OpenEphemeralTab(
    const GURL& url,
    const std::u16string& title,
    content::WebContents& merchant_web_contents) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_PaymentsWindowBridge_openEphemeralTab(
      env, java_payments_window_bridge_, url, title, &merchant_web_contents);
}

void PaymentsWindowBridge::CloseEphemeralTab() {
  Java_PaymentsWindowBridge_closeEphemeralTab(
      base::android::AttachCurrentThread(), java_payments_window_bridge_);
}

void PaymentsWindowBridge::OnNavigationFinished(const GURL& clicked_url) {
  payments_window_delegate_->OnDidFinishNavigationForBnpl(clicked_url);
}

void PaymentsWindowBridge::OnWebContentsObservationStarted(
    content::WebContents* web_contents) {
  if (web_contents) {
    payments_window_delegate_->OnWebContentsObservationStarted(*web_contents);
  }
}

void PaymentsWindowBridge::OnWebContentsDestroyed() {
  payments_window_delegate_->WebContentsDestroyed();
}

void PaymentsWindowBridge::OnUserDeniedTabOpening() {
  payments_window_delegate_->OnUserDeniedTabOpening();
}

}  // namespace autofill::payments

DEFINE_JNI(PaymentsWindowBridge)
